File: src/main/scala/cub/hackerrank/tutorials/a30DaysOfCode/day29/day29.py
#!/usr/bin/env python3

if __name__ == '__main__':
    t = int(input())

    for t_itr in range(t):
        nk = input().split()
        n = int(nk[0])
        k = int(nk[1])

        a = k - 1
        b = 0
        while a > 0:
            lowestZero = 0
            tmp = a
            while tmp > 0 and (tmp & 1) == 1:
                tmp >>= 1
                lowestZero += 1
            b = a + (1 << lowestZero)
            if b <= n:
                break
            a -= 1
        print(a & b)

File: src/main/java/cub/hackerrank/tutorials/a30DaysOfCode/day29/day29.pl
#!/usr/bin/env perl

use strict;
use warnings;

my $t = <>;
$t =~ s/\s+$//;

for (my $t_itr = 0; $t_itr < $t; $t_itr++) {
  my $nk = <>;
  $nk =~ s/\s+$//;
  my @nk = split /\s+/, $nk;

  my $n = $nk[0];
  $n =~ s/\s+$//;

  my $k = $nk[1];
  $k =~ s/\s+$//;

  my $a = $k - 1;
  my $b = 0;
  for (; $a > 0; $a--) {
    my $lowestZero = 0;
    my $tmp = $a;
    while ($tmp > 0 && ($tmp & 1) == 1) {
      $tmp >>= 1;
      $lowestZero++;
    }
    $b = $a + (1 << $lowestZero);
    if ($b <= $n) {
      last;
    }
  }
  print($a & $b, "\n");
}
